### ISSUE TITLE
Dependency on undocumented (or poorly documented) socket.io behaviour

### ISSUE THREAD

**Issue by buu700** [andyet/SimpleWebRTC#441]
Not sure what the best solution here is, but the current implementation of `SimpleWebRTC.joinRoom` makes the current README effectively useless for helping one write a custom `Connection` implementation.

When I started digging into the SimpleWebRTC code to debug my (inexplicably non-working) implementation, seeing that `Connection.emit` was somehow supposed to accept and pass some sort of data into a callback was baffling, as nothing in the docs had indicated anything of the sort. To make things even more confusing, nothing in socket.io's documentation suggested the existence of such behaviour either. It was only after decently extensive investigation within both codebases that I was able to figure out what was going on:

1. `emit('join', ...` seems to be an undocumented alias for socket.io's `join` method.

2. The socket.io server does indeed send information back to the client calling `join`.

While [my workaround](https://github.com/cyph/cyph/blob/next/shared/js/cyph/p2p/p2p.ts#L213) ultimately turned out to be trivial, it shouldn't have required this much effort on my part to validate that what I did there was correct and unlikely to lead to unexpected breakage in edge cases. I'm also not sure how much sense it makes to even offer the ability to specify a custom `Connection` if the code consuming the `Connection` object is relying on specific internal details of `SocketIoConnection` like that.

One possible fix that comes to mind is to get rid of the `connection.emit('join', ...` call (unless there's a specific reason for that style), replace it with `connection.join`, and document a spec for `Connection.join` that the consuming code will continue to respect even in the event that socket.io's API changes (in which case `SocketIoConnection` would need to be updated to adapt, rather than simplewebrtc.js and every project that depends on SimpleWebRTC and a custom `Connection` implementation).

(Amazing work overall though! SimpleWebRTC helped remove a lot of complexity from our voice/video calling code.)

**Comment by xdumaine** [comment]
`emit` is "undocumented" on socket.io's docs because socket.io's docs are quite threadbare. They don't document much above what is basic. You can find more info through discussion boards, slack, and stackoverflow. I too, would appreciate this being better, but I haven't contributed time to improve their docs. 

`emit` as it pertains to socket.io is described here - http://stackoverflow.com/questions/11498508/socket-emit-vs-socket-send, but it's just the standard node event emitter function -https://nodejs.org/api/events.html#events_emitter_emit_eventname_arg1_arg2

The custom Connection model doesn't rely on specific internals of socket.io. Any signaling channel can be used, socket.io or not. The reason you had to dig to understand that the callback should be called is not that SimpleWebRTC is using any reliance on undocumented features of socket.io, but rather that the signalmaster has a behavior (invoking a callback provided in the join event) that wasn't obvious. I think your workaround is straightforward, and all it would take to make that more obvious is basically putting your wrapper into the custom connection code and docs.

The reason `connection.emit('join', name, callback)` is used instead of `connection.join(name)` is that the SimpleWebRTC/Signalmaster rooms have more needs than just basic socket.io rooms.

Thanks for the compliment at the end and for writing this up, but as for "it shouldn't have required this much effort on my part" and the rest of your harsh comments, these were quite unnecessary. This is a free open source project maintained by people in their spare time. The truth is that you were rude and entitled here. The people who've contributed to this library have put in effort so you don't have to. It's not our obligation to completely eliminate effort on your part, and having that attitude is caustic to open source.

One other note: it's probably a good thing that you had to dig through the code to understand how it works. Since you're using this to build a platform based on security, I'd be concerned if you *didn't* understand the workings of open source code that you're using.

**Comment by buu700** [comment]
> Thanks for ... attitude is caustic to open source.

Sorry if something I wrote rubbed you the wrong way, but nothing I said was intended as rude at all. The comment you quoted was meant to be taken literally as constructive feedback, not as an attack on the project or anyone involved with it.

I'm honestly not seeing how you interpreted this report as hostility on my part, and the accusation of entitlement is a bit silly considering I'd just sent a PR for something unrelated (and would be perfectly willing to send one for this issue after getting a consensus from the project maintainers on requirements to move forward with).

> One other note: ... that you're using.

It doesn't hurt (I skim the code of any third-party libs we use as a sanity check anyway, and they're all in scope for audits), but nothing I did here really exactly qualifies as a useful security assessment. After doing some more testing of our new SimpleWebRTC-based code to make sure it works reliably (it isn't in prod yet), I'm planning to submit a new ticket suggesting an [audit from Open Tech Fund](https://www.opentech.fund/lab/red-team-lab) (which I would be happy to personally pursue on behalf of the team).

Anyway, back on topic.

Thanks for the clarification on socket.io. Is it generally understood then that the interfaces they currently expose, even when not sufficiently documented, won't be subject to frequent or unnecessary breaking changes without some sort of communication to their users? If that's just the way they do things, it doesn't sound like much can be done here about the dependency on an "undocumented" feature.

Your other point makes sense to me; sounds like this is a documentation issue more than anything. I might still suggest splitting off a separate `Connection.join` method and having `SocketIoConnection.join` call `this.emit('join', ...` rather than documenting that `Connection.emit` implementations are supposed to be written as `if (event === 'join') { ... } else { ... }`, but that's mostly nitpicking.

**Comment by fippo** [comment]
the lack of documentation of socket.io and its protocol is actually one of the things that makes me greatly prefer [jingle.js](https://github.com/otalk/jingle.js). I feel your pain but I don't see anyone interested in and at the same time capabable of writing a complete specification + implementation that is as good as the jingle specs (h/t @stpeter)
